vcenter_operator/phelm.py

```diff
@@ -12,36 +11,137 @@
-        template_names = env.list_templates(
-            filter_func=lambda x: (x.startswith(scope)
-                                   and x.endswith('.yaml.j2')))
+    def render(self, scope, options, service_users, vcenter_service_user_tracker):
+        template_names = env.list_templates(filter_func=lambda x: (x.startswith(scope) and x.endswith(".yaml.j2")))
+        service_user_crds = vcenter_service_user_crd_loader.get_mapping()
         for template_name in template_names:
             try:
                 template = env.get_template(template_name)
                 # e.g. vcenter_cluster/namespace/cr-name
                 namespace = template_name.split('/')[1]
-                result = template.render(options)
+                result = self._inject_service_user_info_and_render(
+                    template, service_users, vcenter_service_user_tracker, service_user_crds, options
+                )
                 owner = env.get_source_owner(template_name)
                 self.add(result, owner, namespace)
             except (TemplateError, YAMLError):
                 LOG.exception("Failed to render %s", template_name)
+        # Order the item by kind to ensure that Secrets are created before ConfigMaps and Deployments
+        self.order_items()
+
+    def _inject_service_user_info_and_render(
+        self, template, service_users, vcenter_service_user_tracker, service_user_crds, options
+    ):
+        """Check if template uses service-user and inject necessary information into the template"""
+
+        # Check template source for service-user variable
+        if not hasattr(template, "environment") or not hasattr(template.environment, "loader"):
+            LOG.debug("Template does not have an environment or loader")
+            return template.render(options)
+
+        template_source_list = template.environment.loader.get_source(template.environment, template.name)
+        if not template_source_list:
+            LOG.warning("Template source not found")
+            return template.render(options)
+
+        template_source = template_source_list[0]
+
+        match = self.SERVICE_USER_MATCH.search(template_source)
+        if not match:
+            return template.render(options)
+
+        service_name = match.group(1)
+        service_user_path = f"{options['region']}/vcenter-operator/{service_name}/{options['vcenter_name']}"
+
+        if service_name not in service_user_crds.keys():
+            LOG.warning(f"Service-user {service_name} not found - skipping this host")
+            raise ServiceUserNotFoundError()
+
+        if service_user_path not in service_users:
+            LOG.warning(f"Service-user path {service_name} not found - skipping this host")
+            raise ServiceUserNotFoundError()
+
+        # Separate the secret and deployment rendering to avoid rendering a secret in the deployment accidentally in
+        # the future
+        if "kind: Secret" in template_source:
```

Comment:
Do we have to rely on string-checking here? Is the `kind` only available after the rendering?